akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -247,3 +224,78 @@ private[akka] final class HeadOptionStage[T] extends SinkStage[T, Future[Option[
     }, p.future)
   }
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] class QueueSink[T]() extends SinkStage[T, SinkQueue[T]]("queueSink") {
+  trait RequestElementCallback {
+    def requestElement(p: Promise[Option[T]]): Unit
+  }
+
+  type Received[E] = Try[Option[E]]
+
+  val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+  require(maxBuffer > 0, "Buffer size must be greater than 0")
+  val requestInProgress = new AtomicBoolean()
+  val buffer = new LinkedBlockingQueue[Received[T]](maxBuffer + 1)
```

Comment:
Please do not use blocking as a means of conveying back-pressure: wrapping `take` in a Future is not a nice thing to do.